package uk.gov.ons.census.notifysvc.config;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.ObjectMapper;
import uk.gov.ons.census.notifysvc.utils.ObjectMapperFactory;

@Configuration
public class NotifyConfiguration {

  @Value("${notifyserviceconfigfile}")
  private String configFile;

  public static final ObjectMapper OBJECT_MAPPER = ObjectMapperFactory.objectMapper();

  @Bean
  public NotifyServiceRefMapping notifyServiceRefMapping() {
    Map<String, Map<String, String>> rawJsonConfig;

    try (InputStream configFileStream = new FileInputStream(configFile)) {
      rawJsonConfig = OBJECT_MAPPER.readValue(configFileStream, Map.class);
    } catch (JacksonException | FileNotFoundException e) {
      throw new RuntimeException(e);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }

    NotifyServiceRefMapping notifyServiceRefMapping = new NotifyServiceRefMapping();

    for (String notifyServiceRef : rawJsonConfig.keySet()) {
      notifyServiceRefMapping.addNotifyClient(
          notifyServiceRef,
          rawJsonConfig.get(notifyServiceRef).get("base-url"),
          rawJsonConfig.get(notifyServiceRef).get("api-key"),
          rawJsonConfig.get(notifyServiceRef).get("sender-id"));
    }

    return notifyServiceRefMapping;
  }
}
